Blinking cursor + text editing improvements

futurepaul opened the pull request:
If this is too messy I can strip out everything but the blinking, but thought I'd run it by you like this first. This version needs `unicode-segmentation` as a dependency, so that's also worth discussing.

Also, the get_layout function is copied from Label originally, so if we end up with [caching in Label](https://github.com/xi-editor/druid/pull/116) it would be good to pull that in here as well.

raphlinus commented on druid-shell/src/lib.rs:
```diff
@@ -17,6 +17,8 @@
 pub use piet_common as piet;
 pub use piet_common::kurbo;
 
+pub use unicode_segmentation;
```
Hmm. I'm not sure I want druid-shell to become a "batteries included" aggregator for libraries we want. It feels like this dependency is better in druid than druid-shell. We're doing this for kurbo and piet because those are necessary dependencies from druid-shell.

That said, the idea that druid depends on unicode-segmentation seems quite reasonable to me.

raphlinus commented on src/widget/textbox.rs:
```diff
@@ -35,30 +44,131 @@ const BORDER_WIDTH: f64 = 1.;
+    fn insert(&mut self, src: &mut String, new: &str) {
+        // TODO: handle incomplete graphemes
+        let (left, right) = (self.selection.min(), self.selection.max());
+        src.replace_range(left..right, new);
```
TIL this method! It's great, such an improvement over doing truncation and push of slices by hand.

raphlinus commented on src/widget/textbox.rs:
```diff
@@ -35,30 +44,131 @@ const BORDER_WIDTH: f64 = 1.;
+    fn backspace(&mut self, src: &mut String) {
+        if self.selection.is_caret() {
+            let cursor = self.cursor();
+            let new_cursor = prev_grapheme(&src, cursor);
```
I'd put in a TODO that grapheme isn't really the right unit. FYI, I consider https://github.com/xi-editor/xi-editor/blob/master/rust/core-lib/src/backspace.rs to be pretty much the authoritative source of what this logic should be.

raphlinus commented on src/widget/textbox.rs:
```diff
@@ -90,30 +199,63 @@ impl Widget<String> for TextBox {
-                    let offset = text_layout.width() - self.width + (PADDING_LEFT * 2.) + 1.;
-                    rc.transform(Affine::translate(Vec2::new(-offset, 0.)));
+                let padding = PADDING_LEFT * 2.;
+                if cursor_x > self.width + self.hscroll_offset - padding {
```
In *general* I like paint to be a more or less pure function that just paints the view; this logic feels like it should be split out and invoked from anything that causes a change (event, update). But that's more of a feeling; paint is in fact allowed to do stuff like this.

futurepaul replied:
That's a good rule of thumb. I moved it to a method and am calling from event now. In doing so, I realized that there were definitely times I was doing this calculation that I didn't need to be doing it!

raphlinus commented on src/widget/textbox.rs:
```diff
@@ -90,30 +199,63 @@ impl Widget<String> for TextBox {
+                    self.hscroll_offset = cursor_x
                 }
+
+                rc.transform(Affine::translate(Vec2::new(-self.hscroll_offset, 0.)));
```
Minor style point, the `Vec2::new` can be elided because the translate constructor takes its argument by `Into<Vec2>`.

raphlinus commented on src/widget/textbox.rs:
```diff
@@ -144,18 +289,102 @@ impl Widget<String> for TextBox {
                 match key_event {
+                    // Copy (Ctrl+C || Cmd+C)
+                    event
+                        if (event.mods.meta || event.mods.ctrl)
```
Note that `druid_shell::keycodes::command_modifier()` tells you which modifier is the command modifier for the platform. It's not available in easy-to-use form though. Also see #81 for related ideas.

This question of keybindings etc needs attention (including the fact that a lot of these are likely best expressed as shortcuts available on an "Edit" menu), but I'm happy getting something useful in early.

futurepaul replied:
I couldn't figure out how to compare a `KeyModifier` with a `Modifier`. Would it be okay to do like this?

```rust
const fn command_modifier() -> KeyModifiers {
    #[cfg(target_os = "macos")]
    {
        KeyModifiers {
            shift: false,
            alt: false,
            ctrl: false,
            meta: true,
        }
}
    #[cfg(not(target_os = "macos"))]
    {
        KeyModifiers {
            shift: false,
            alt: false,
            ctrl: false,
            meta: true,
        }
    }
}
```

cmyr replied:
mm, this is interesting.

I think that the correct way to handle this is to have some sort of `CommandList` struct where you register the commands you're looking for, which can have per-platform overrides, and then you pass events to that struct to determine if a command has occurred, but I'll need to think about that a bit more.

For the time being, I think adding `fn contains_sys_command_modifier(self) -> bool` to `KeyModifier` is... okay. 🤢

cmyr commented on src/widget/textbox.rs:
```diff
@@ -35,30 +44,153 @@ const BORDER_WIDTH: f64 = 1.;
+
+    fn insert(&mut self, src: &mut String, new: &str) {
+        // TODO: handle incomplete graphemes
+        let (left, right) = (self.selection.min(), self.selection.max());
```
Non-blocking: you could have a fn `Selection::range() -> Range<usize>`, if you felt like it.

futurepaul replied:
I like it.

cmyr commented on src/widget/textbox.rs:
```diff
@@ -35,30 +44,153 @@ const BORDER_WIDTH: f64 = 1.;
+        }
+    }
+
+    fn copy_text(&self, input: String) {
```
maybe throw in a `TODO` note here, explaining you're waiting for copy/past to get implemented?

futurepaul commented:
Alright, I believe I've addressed everything. I also added a few more doc strings. Since I'm not quite following your ambitions for event matching ergonomics, I think `fn contains_sys_command_modifier(self) -> bool` looks great! :)

cmyr commented:
Feel like I had a bit of a brain fart, since `contains_sys_command_modifier` will also return true for multi-modifier combos, which we might not want. This might be okay for now, but has clear limitations longer-term. I'm going to sketch out some command handling stuff today, and see if we can't improve that situation.

futurepaul commented:
Okay, rebased to #121. I think it went well. Because of the `matches` method instead of PartialEq I figured it would be best to use HotKeys throughout, even where it isn't strictly required, so that nobody gets confused with `==`.

Also I caught an edgecase with hscroll that I was missing! As simple as hscroll logic looks, it's really been a brain challenge for me to do it correctly.

cmyr commented:
okay, cool.

This makes me wonder if I should actually have a fn `matches` on `KeyEvent`, which takes an `impl Into<HotKey>`; this would let us just have code like,

```rust

match event {
    k_e if k_e.matches((SysMod::Cmd, "c")) => (),
    k_e if k_e.matches((None, KeyCode::LeftArrow)) => (),
}
```

Although this might require a bunch of this stuff to go into druid-shell. That might make sense anyway, though...

cmyr commented on src/widget/textbox.rs:
```diff
@@ -14,18 +14,28 @@
+
+use crate::unicode_segmentation::GraphemeCursor;
 
 const BACKGROUND_GREY_LIGHT: Color = Color::rgba8(0x3a, 0x3a, 0x3a, 0xff);
```
one last thought: since env has landed, we can start moving these sorts of things into that, if we like? But I'm happy for this to be followup.

futurepaul replied:
I was planning on a styled widgets PR next, so I think I'd prefer to merge now.

cmyr replied:
sounds good!
